integration/querydsl/blaze-persistence-querydsl-expressions/src/main/java/com/blazebit/persistence/querydsl/BlazeCriteriaBuilderRenderer.java

```diff
@@ -0,0 +1,1308 @@
+        if (expr instanceof FactoryExpression) {
+            return ((FactoryExpression<?>) expr).getArgs();
+        } else {
+            return Collections.singletonList(expr);
+        }
+    }
+
+    private JoinType getJoinType(JoinExpression joinExpression) {
+        switch (joinExpression.getType()) {
+            case INNERJOIN:
+            case JOIN:
+                return JoinType.INNER;
+            case LEFTJOIN:
+                return JoinType.LEFT;
+            case RIGHTJOIN:
+                return JoinType.RIGHT;
+            case FULLJOIN:
+                return JoinType.FULL;
+            default:
+                throw new IllegalArgumentException("Join has no equivalent JoinType");
+        }
+    }
+
+    private void renderOrderSpecifier(OrderSpecifier<?> orderSpecifier, OrderByBuilder<?> criteriaBuilder) {
+        String orderExpression = renderExpression(orderSpecifier.getTarget());
+        boolean ascending = orderSpecifier.isAscending();
+        switch (orderSpecifier.getNullHandling()) {
+            case Default:
+                criteriaBuilder.orderBy(orderExpression, ascending);
+                break;
+            case NullsFirst:
+                criteriaBuilder.orderBy(orderExpression, ascending, true);
+                break;
+            case NullsLast:
+                criteriaBuilder.orderBy(orderExpression, ascending, false);
+                break;
+            default:
+                throw new IllegalArgumentException("Null handling not implemented for " + orderSpecifier.getNullHandling());
+        }
+    }
+
+    private void renderSingleSelect(Expression<?> select, final SelectBuilder<?> selectBuilder) {
+        String alias = null;
+
+        if (select instanceof Operation<?>) {
+            Operation<?> operation = (Operation<?>) select;
+            if (operation.getOperator() == Ops.ALIAS) {
+                select = operation.getArg(0);
+                alias = ((Path<?>) operation.getArg(1)).getMetadata().getName();
+            }
+        }
+
+        final String finalAlias = alias;
+        setExpressionSubqueries(select,
+                new BlazeCriteriaBuilderRenderer.Function<String, Object>() {
+                    @Override
+                    public Object apply(String expression) {
+                        return finalAlias != null ? selectBuilder.select(expression, finalAlias) : selectBuilder.select(expression);
+                    }
+                },
+                new BlazeCriteriaBuilderRenderer.Function<String, MultipleSubqueryInitiator<?>>() {
+                    @Override
+                    public MultipleSubqueryInitiator<?> apply(String expression) {
+                        return finalAlias != null ? selectBuilder.selectSubqueries(expression, finalAlias) : selectBuilder.selectSubqueries(expression);
+                    }
+                });
+    }
+
+    private String renderExpression(Expression<?> select) {
```

Comment:
The `JPQLNextSerializer` is able to serialize to Strings, i.e.

```java
JPQLNextSerializer jpqlNextSerializer = new JPQLNextSerializer();
expression.accept(jpqlNextSerializer, null);
String output = jpqlNextSerializer.toString();
```

Unfortunately, the `JPQLSerializer` that it's based on treats outer queries slightly different. It probably has to. Outer queries, orthogonally to our FullQueryBuilder, just support features plain subqueries don't.

I didn't bother too much with it, because there is no full query parser for JPQL.Next, so I have to render against the CriteriaBuilder API anyway. The desired way to "serialize" a query, and the one I use in `AbstractBlazeJPAQuery`, is IMO to just render it and invoke `getQueryString()` on the `Queryable`.

Another complication here is that the `BlazeCriteriaBuilderRenderer` has to hook directly on the `JPQLSerializer`. For example, it renders subqueries as an alias, which is not really practical for to-string serialization.

Reply:
So is the JPQLNextSerializer just for the _human readable_ string representation?
Anyway, I was thinking that moving this method to the serializer class would allow making `WindowContainerBuilderDefaultVisitorImpl` static and let it accept just the serializer, but I probably missed some context usage.

Reply:
The JPQLNextSerializer just makes the required changes to the JPQLSerializer to deal with set operations and work around some issues in QueryDSL (for example how they use HQL casts).

It generally produces executable queries, but there are some quirks when it gets to CTEs, because the bindings are not "expanded" until the query is actually rendered in the renderer.

Furthermore it may produce some JPQLNext specific dialect such as window clauses for which no full query parser yet exists.

All that magic is handled in the renderer instead.